File: clojure/9.clj
(defn pack [lst]
  (if (empty? lst)
    [[]]
    (if (= (count lst) 1)
      [lst]
      (let [packtail (pack (rest lst))]
        (if (= (first lst) (first (rest lst)))
          (cons (cons (first lst)
                      (first packtail))
                (rest packtail))
          (cons [(first lst)] packtail))))))

File: scala/9.scala
object Solution9 {
  def pack[T](lst: List[T]): List[List[T]] = {
    lst match {
      case List() => List(List())
      case List(h) => List(List(h))
      case x::y::_ => {
        val packtail = pack(lst.tail)
        if (x == y) (x :: packtail.head) :: packtail.tail
        else List(x) :: packtail
      }
    }
  }
}
